### PR TITLE
reimplement merge to drop mergo dependency

### LINKED ISSUE #107499: remove use of github.com/imdario/mergo (currently use is incompatible with latest mergo version, latest mergo version introduces bugs)
### What happened?

client-go behavior changes depending on version of github.com/imdario/mergo used. Kubernetes pins to `v0.3.5`. Many other projects in the ecosystem, including Kubernetes projects like sigs.k8s.io/controller-runtime, use newer versions like `0.3.12`.

If a consumer of client-go is not careful, they will very likely automatically use `mergo`. This will subtly give different behavior than client-go declares. Users of client-go must therefor use a `replace` for mergo (if they care about this issue), which also implies all downstream consumers must do the same and `go install` will not work.

### What did you expect to happen?

client-go behavior is the same regardless of mergo version.

### How can we reproduce it (as minimally and precisely as possible)?

```go
package kube

import (
	"fmt"
	"os"
	"path/filepath"
	"testing"

	"k8s.io/client-go/tools/clientcmd"
)

func TestMergo(t *testing.T) {
	config1, err := mkKubeConfig("1.1.1.1", "3.3.3.3")
	if err != nil {
		t.Fatalf("Failed to create a sample kubernetes config file. Err: %v", err)
	}
	defer os.RemoveAll(filepath.Dir(config1))
	config2, err := mkKubeConfig("2.2.2.2", "4.4.4.4")
	if err != nil {
		t.Fatalf("Failed to create a sample kubernetes config file. Err: %v", err)
	}
	defer os.RemoveAll(filepath.Dir(config2))

	if err := os.Setenv("KUBECONFIG", config1+":"+config2); err != nil {
		t.Fatalf("Failed to set KUBECONFIG environment variable")
	}

	loadingRules := clientcmd.NewDefaultClientConfigLoadingRules()
	loadingRules.DefaultClientConfig = &clientcmd.DefaultClientConfig
	configOverrides := &clientcmd.ConfigOverrides{
		ClusterDefaults: clientcmd.ClusterDefaults,
	}

	//resp, err := BuildClientConfig("", "")
	resp, err := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(loadingRules, configOverrides).ClientConfig()
	if err != nil {
		t.Fatal(err)
	}
	wantHost := "https://1.1.1.1:8001"
	if resp != nil && resp.Host != wantHost {
		t.Fatalf("Incorrect host. Got: %s, Want: %s", resp.Host, wantHost)
	}
}

func mkKubeConfig(cluster1Host string, cluster2Host string) (string, error) {
	tempDir, err := os.MkdirTemp("/tmp/", ".kube")
	if err != nil {
		return "", err
	}
	filePath := filepath.Join(tempDir, "config")

	template := `apiVersion: v1
kind: Config
clusters:
- cluster:
    insecure-skip-tls-verify: true
    server: https://%s:8001
  name: cluster.local
- cluster:
    insecure-skip-tls-verify: true
    server: https://%s:8001
  name: cluster2.local
contexts:
- context:
    cluster: cluster.local
    namespace: default
    user: admin
  name: cluster.local-context
- context:
    cluster: cluster2.local
    namespace: default
    user: admin
  name: cluster2.local-context
current-context: cluster.local-context
preferences: {}
users:
- name: admin
  user:
    token: sdsddsd`

	sampleConfig := fmt.Sprintf(template, cluster1Host, cluster2Host)
	err = os.WriteFile(filePath, []byte(sampleConfig), 0o644)
	if err != nil {
		return "", err
	}
	return filePath, nil
}
```

Run with v0.3.5:
```
=== RUN   TestMergo
--- PASS: TestMergo (0.00s)
PASS
```

Run with v0.3.12:
```
=== RUN   TestMergo
    repro_test.go:41: Incorrect host. Got: https://2.2.2.2:8001, Want: https://1.1.1.1:8001
--- FAIL: TestMergo (0.00s)

FAIL
```

If its easier you can also remove the `replace` in http://github.com/istio/istio and run `go test ./pkg/kube -run TestBuildClientConfig` which will give the same result. The above test is just a simplified version of this test.

### Anything else we need to know?

_No response_

### Kubernetes version

<details>
v0.23.1, but this has been an issue for almost 2 years
</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by liggitt**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Reimplements the narrow slice of merge capability we need from mergo to allow dropping a dependency.

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/107499

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

/area code-organization
/sig architecture cli api-machinery
/cc @deads2k @dims

### CODE REVIEW

File: staging/src/k8s.io/client-go/tools/clientcmd/merge.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+	"fmt"
+	"reflect"
+	"strings"
+)
+
+// recursively merges src into dst:
+// - non-pointer struct fields are recursively merged
+// - maps are shallow merged with src keys taking priority over dst
+// - non-zero src fields encountered during recursion that are not maps or structs overwrite and recursion stops
+func merge[T any](dst, src *T) error {
```
Comment (dims) [inline staging/src/k8s.io/client-go/tools/clientcmd/merge.go:29]: looks like this is equivalent to the `mergo.Merge` with `WithOverride` option

**Comment by dims** [comment]
@liggitt i was trying some of the test cases in the original mergo repo with the specific tag and we seem to trip up this one. can you please peek?

https://github.com/darccio/mergo/blame/v0.3.6/issue64_test.go#L19

```
{Student{"Jack", []string{}}, Student{"Tom", []string{"1"}}, []string{"1"}}
```

**Comment by dims** [comment]
@liggitt also i don't know if we end up merging timestamps in our use case... `TestIssue38MergeWithOverwrite` and `TestOverwriteZeroSrcTime` are wonky as well in addition to the one above.

**Comment by dims** [comment]
@liggitt can we do this for this release?

File: staging/src/k8s.io/client-go/tools/clientcmd/merge.go (hunk 1 above)
Comment (liggitt) [reply staging/src/k8s.io/client-go/tools/clientcmd/merge.go:30]: yes, that was our only use

**Comment by liggitt** [comment]
Rebased this, pulled in a tmp commit that runs all the mergo unit tests that exercise WithOverwrite invocations (including TestIssue38MergeWithOverwrite, TestOverwriteZeroSrcTime, and TestIssue64MergeSliceWithOverride, as noted above).

Reworked the implementation to make all relevant mergo unit tests pass as well.

Once this has review, I'll drop the TMP commit with those unit tests.

**Comment by liggitt** [comment]
(lint jobs are complaining about the mergo unit test files that will be dropped)

**Comment by dims** [comment]
Looks good to me @liggitt please drop the `TMP` commit and i'll add the `/lgtm`. thanks for testing with the mergo tests!

**Comment by liggitt** [comment]
For posterity, clean unit test run at https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/126764/pull-kubernetes-unit/1839831403240361984

**Comment by liggitt** [comment]
done

**Comment by dims** [comment]
some linter and verify-vendor errors ... looks legit

**Comment by liggitt** [comment]
```
elseif: can replace 'else {if cond {}}' with 'else if cond {}' (gocritic)
		} else {
```

🙄 the linter is *quite* irritating at times

**Pull request merged**